src/client/gui/lib/settings/hotkey.dart

```diff
@@ -0,0 +1,199 @@
+  final SingleActivator? value;
+  final ValueChanged<SingleActivator?>? onSave;
+
+  const HotkeyRecorder({super.key, this.value, this.onSave});
+
+  @override
+  State<HotkeyRecorder> createState() => HotkeyRecorderState();
+}
+
+class HotkeyRecorderState extends State<HotkeyRecorder> {
+  final focusNode = FocusNode();
+  late var hasFocus = focusNode.hasFocus;
+  var alt = false;
+  var control = false;
+  var meta = false;
+  var shift = false;
+  LogicalKeyboardKey? key;
+  var shouldSave = false;
+
+  @override
+  void initState() {
+    super.initState();
+    set(widget.value);
+  }
+
+  @override
+  void dispose() {
+    focusNode.dispose();
+    super.dispose();
+  }
+
+  void set(SingleActivator? value) {
+    alt = value?.alt ?? false;
+    control = value?.control ?? false;
+    meta = value?.meta ?? false;
+    shift = value?.shift ?? false;
+    key = value?.trigger;
+  }
+
+  void setAlt(bool pressed) => alt = pressed;
+
+  void setCtrl(bool pressed) => control = pressed;
+
+  void setMeta(bool pressed) => meta = pressed;
+
+  void setShift(bool pressed) => shift = pressed;
+
+  late final modifierSetters = {
+    LogicalKeyboardKey.altLeft: setAlt,
+    LogicalKeyboardKey.altRight: setAlt,
+    LogicalKeyboardKey.controlLeft: setCtrl,
+    LogicalKeyboardKey.controlRight: setCtrl,
+    LogicalKeyboardKey.metaLeft: setMeta,
+    LogicalKeyboardKey.metaRight: setMeta,
+    LogicalKeyboardKey.superKey: setMeta,
+    LogicalKeyboardKey.shiftLeft: setShift,
+    LogicalKeyboardKey.shiftRight: setShift,
+  };
+
+  KeyEventResult handleKeyEvent(_, KeyEvent event) {
+    if (event is KeyRepeatEvent) return KeyEventResult.handled;
+
+    shouldSave = false;
+    final pressed = event is KeyDownEvent;
+    final logicalKey = event.logicalKey;
+
+    final modifierSetter = modifierSetters[logicalKey];
+    if (modifierSetter != null) {
+      setState(() => modifierSetter.call(pressed));
+    } else {
+      setState(() => key = pressed ? logicalKey : null);
+    }
+
+    if (key != null && [alt, control, meta].any((pressed) => pressed)) {
+      shouldSave = true;
+      focusNode.unfocus();
+    }
+
+    return KeyEventResult.handled;
+  }
+
+  void handleFocusChange(bool focused) {
+    setState(() {
+      hasFocus = focused;
+      if (focused) {
+        set(null);
+        shouldSave = true;
+      } else if (shouldSave) {
+        shouldSave = false;
+        final trigger = key;
+        final activator = trigger != null
+            ? SingleActivator(
+                trigger,
+                alt: alt,
+                control: control,
+                meta: meta,
+                shift: shift,
+                includeRepeats: false,
+              )
+            : null;
+        widget.onSave?.call(activator);
+      } else {
+        set(widget.value);
+      }
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final keyLabel = key?.keyLabel ?? '...';
+    final modifiers = [
```

Comment:
Tiny nitpick: I believe the order of the keys in the shortcuts is usually Ctrl > Alt > Shift > Super > rest. For instance, I would expect to see Ctrl+Alt+U rather than Alt+Ctrl+U. But feel free to leave this for later.

Reply:
That is a trivial fix, so I'll do it.